File: src/test/java/com/ohgnarly/fileripper/services/FlatFileServiceTest.java
package com.ohgnarly.fileripper.services;

import com.ohgnarly.fileripper.exceptions.FileRipperException;
import com.ohgnarly.fileripper.models.FileDefinition;
import org.junit.Test;

import java.io.File;

import static com.ohgnarly.fileripper.integration.utilities.DataUtility.*;

public class FlatFileServiceTest {
    private FlatFileService flatFileService;

    @Test(expected = FileRipperException.class)
    public void testProcessFile_GivenDelimitedFile_AndInvalidFileFormat_ShouldThrowIOException() throws Throwable {
        //arrange
        FileDefinition fileDefinition = createDelimitedFileDefinition("|");
        fileDefinition.getFieldDefinitions().add(createFieldDefinition("address", null, null));
        File file = createDelimitedFile("|", false);
        flatFileService = new FlatFileService(fileDefinition);

        //act
        flatFileService.processFile(file);
    }

    @Test(expected = FileRipperException.class)
    public void testProcessFile_GivenFixedFile_AndLastFieldTooLong_ShouldThrowIOException() throws Throwable {
        //arrange
        FileDefinition fileDefinition = createFixedFileDefinition();
        fileDefinition.getFieldDefinitions().get(2).setFieldLength(11);
        File file = createFixedFile(true);
        flatFileService = new FlatFileService(fileDefinition);

        //act
        flatFileService.processFile(file);
    }

    @Test(expected = FileRipperException.class)
    public void testProcessFile_GivenTooManyFields_ShouldThrowIOException() throws Throwable {
        //arrange
        FileDefinition fileDefinition = createFixedFileDefinition();
        fileDefinition.getFieldDefinitions().add(createFieldDefinition("address", 36, 0));
        File file = createFixedFile(true);
        flatFileService = new FlatFileService(fileDefinition);

        //act
        flatFileService.processFile(file);
    }
}

File: src/test/java/com/ohgnarly/fileripper/services/FlatFileServiceTest.kt
package com.ohgnarly.fileripper.services

import com.ohgnarly.fileripper.exceptions.FileRipperException
import com.ohgnarly.fileripper.testhelpers.*
import org.junit.Test
import kotlin.test.assertEquals

class FlatFileServiceTest {
    private var flatFileService: FlatFileService? = null

    @Test
    fun testProcessFile_GivenDelimitedFile_ReturnsFileOutput() {
        //arrange
        val file = buildDelimitedFile("|", false)
        flatFileService = FlatFileService(buildDelimitedFileDefinition("|"))

        //act
        val fileOutput = flatFileService!!.processFile(file)

        //assert
        assertEquals(file.name, fileOutput.fileName)
        assertFileRecords(fileOutput.records)
    }

    @Test
    fun testProcessFile_GivenFixedWidthFile_ReturnsFileOutput() {
        //arrange
        val file = buildFixedFile(true)
        flatFileService = FlatFileService(buildFixedFileDefinition())

        //act
        val fileOutput = flatFileService!!.processFile(file)

        //assert
        assertEquals(file.name, fileOutput.fileName)
        assertFileRecords(fileOutput.records)
    }

    @Test(expected = FileRipperException::class)
    fun testProcessFile_GivenDelimitedFile_AndInvalidFileFormat_ThrowsFileRipperException() {
        //arrange
        val fileDefinition = buildDelimitedFileDefinition("|")
        fileDefinition.fieldDefinitions.add(buildFieldDefinition("address", null, null))
        flatFileService = FlatFileService(fileDefinition)

        val file = buildDelimitedFile("|", false)

        //act
        flatFileService!!.processFile(file)
    }

    @Test(expected = FileRipperException::class)
    fun testProcessFile_GivenFixedFile_AndLastFieldTooLong_ThrowsFileRipperException() {
        //arrange
        val fileDefinition = buildFixedFileDefinition()
        fileDefinition.fieldDefinitions[2].fieldLength = 11
        flatFileService = FlatFileService(fileDefinition)

        val file = buildFixedFile(true)

        //act
        flatFileService!!.processFile(file)
    }

    @Test(expected = FileRipperException::class)
    fun testProcessFile_GivenTooManyFields_ThrowsFileRipperException() {
        //arrange
        val fileDefinition = buildFixedFileDefinition()
        fileDefinition.fieldDefinitions.add(buildFieldDefinition("address", 36, 0))
        flatFileService = FlatFileService(fileDefinition)

        val file = buildFixedFile(true)

        //act
        flatFileService!!.processFile(file)
    }
}